[WTF] Use `std::trunc` in `isIntegral` function for Correct Integral Check instead of `static_cast<int>`

Gumichocopengin8 opened the pull request:
#### 3b0ec9763829b8fce6e7813b425f03d0ef84c65d
<pre>
[WTF] Use `std::trunc` in `isIntegral` function for Correct Integral Check instead of `static_cast&lt;int&gt;`
<a href="https://bugs.webkit.org/show_bug.cgi?id=292617">https://bugs.webkit.org/show_bug.cgi?id=292617</a>

Reviewed by Chris Dumez.

`static_cast&lt;int&gt;` does not cover the entire range of float, such as `3.4e38f`.
In order to avoid the potential bug,
use `std::trunc` instead of `static_cast&lt;int&gt;` for more reliable check.

* Source/WTF/wtf/MathExtras.h:
(WTF::isIntegral):

Canonical link: <a href="https://commits.webkit.org/294596@main">https://commits.webkit.org/294596@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d37ec1dfe41275ea58ae30b895f18c537e9544e0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/102403 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/22072 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/12386 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/107564 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/53039 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/22380 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/30578 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/77914 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34900 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/105409 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17335 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/92438 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58251 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/17170 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/10466 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/52397 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/95074 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/86991 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10537 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/109939 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/101012 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/29535 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/21776 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/86890 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/29899 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/88634 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/86481 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22001 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31307 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9027 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/23777 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/29463 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/34765 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/124646 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/29274 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34591 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/32597 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/30833 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WTF/wtf/MathExtras.h:
```diff
@@ -494,9 +494,9 @@ inline typename std::enable_if<std::is_floating_point<T>::value, T>::type nanPro
 }
 
-inline bool isIntegral(float value)
+inline constexpr bool isIntegral(const float value)
```
The change looks fine overall but please do not mark the parameter as `const` here. In WebKit, we don't use `const` for parameters passed by value.

Gumichocopengin8 replied:
Fixed

cdumez reviewed (approved):
LGTM

Gumichocopengin8 committed:
[WTF] Use `std::trunc` in `isIntegral` function for Correct Integral Check instead of `static_cast<int>`
https://bugs.webkit.org/show_bug.cgi?id=292617

Reviewed by Chris Dumez.

`static_cast<int>` does not cover the entire range of float, such as `3.4e38f`.
In order to avoid the potential bug,
use `std::trunc` instead of `static_cast<int>` for more reliable check.

* Source/WTF/wtf/MathExtras.h:
(WTF::isIntegral):

Canonical link: https://commits.webkit.org/294596@main

darinadler commented on Source/WTF/wtf/MathExtras.h:
```diff
@@ -499,9 +499,9 @@ inline typename std::enable_if<std::is_floating_point<T>::value, T>::type nanPro
+inline constexpr bool isIntegral(float value)
 {
-    return static_cast<int>(value) == value;
+    return std::trunc(value) == value;
```
This changes behavior and, I believe, regresses correctness for _infinity_ and _nan_; the function now returns _true_ and before this change it returned _false_. Maybe that’s OK given how we use the function.

Gumichocopengin8 replied:
I think both implementations appear to return `false` when given `NaN`. However, as noted, the behavior with `infinity` has changed in this PR.

Fixed the implementation of `isIntegral` to return `false` with when given `infinity` in #45264
Also, added test cases in #45264 to ensure the expected `isIntegral` behavior :)
Thanks for the feedback!

Gumichocopengin8 replied:
Thanks for highlighting this!
